File: src/meta_schedule/space_generator/space_generator.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "../../target/parsers/aprofile.h"
#include "../utils.h"

namespace tvm {
namespace meta_schedule {

String GetRuleKindFromTarget(const Target& target) {
  if (target->kind->name == "llvm") {
    static const PackedFunc* target_has_feature_fn_ptr =
        runtime::Registry::Get("target.target_has_feature");
    ICHECK(target_has_feature_fn_ptr != nullptr)
        << "The `target.target_has_feature` func is not in tvm registry.";
    bool have_avx512vnni = (*target_has_feature_fn_ptr)("avx512vnni", target);
    bool have_avxvnni = (*target_has_feature_fn_ptr)("avxvnni", target);
    if (have_avx512vnni || have_avxvnni) {
      return "vnni";
    } else {
      bool have_avx512f = (*target_has_feature_fn_ptr)("avx512f", target);
      bool have_avx512bw = (*target_has_feature_fn_ptr)("avx512bw", target);
      if (have_avx512bw && have_avx512f) {
        return "avx512";
      }
    }

    TargetJSON target_json = target::parsers::aprofile::ParseTarget(target->Export());
    TargetFeatures afeatures = Downcast<TargetFeatures>(target_json.at("features"));

    if (Downcast<Bool>(afeatures.at("has_dotprod"))) {
      return "dotprod";
    }
    if (Downcast<Bool>(afeatures.at("has_asimd"))) {
      return "asimd";
    }
    return "llvm";
  }
  if (target->kind->name == "hexagon") {
    return "hexagon";
  }
  if (target->kind->name == "cuda") {
    if (Optional<String> opt_sm = target->GetAttr<String>("arch")) {
      std::string sm = opt_sm.value();
      if (support::StartsWith(sm, "sm_")) {
        sm = sm.substr(3);
        try {
          if (std::stoi(sm) >= 75) {
            return "cuda-tensorcore";
          }
        } catch (const std::invalid_argument& e) {
          LOG(WARNING) << "ValueError: Unable to parse `target.arch`: " << sm
                       << ". Details: " << e.what();
        }
      }
    }
    return "cuda";
  }

  if (IsGPUTarget(target->kind->name)) {
    return "cuda";
  }

  if (target->kind->name == "c") {
    return "c";
  }
  LOG(FATAL) << "Unsupported target: " << target;
  throw;
}

void SpaceGeneratorNode::InitializeWithTuneContext(const TuneContext& context) {
  if (context->target.defined() &&  //
      !(sch_rules.defined() &&      //
        postprocs.defined() &&      //
        mutator_probs.defined())) {
    String kind = GetRuleKindFromTarget(context->target.value());
    Array<ScheduleRule> default_sch_rules;
    Array<Postproc> default_postprocs;
    Map<Mutator, FloatImm> default_mutator_probs;
    // for target with skylake-avx512
    if (kind == "llvm") {
      default_sch_rules = ScheduleRule::DefaultLLVM();
      default_postprocs = Postproc::DefaultLLVM();
      default_mutator_probs = Mutator::DefaultLLVM();
    } else if (kind == "cuda") {
      default_sch_rules = ScheduleRule::DefaultCUDA();
      default_postprocs = Postproc::DefaultCUDA();
      default_mutator_probs = Mutator::DefaultCUDA();
    } else if (kind == "cuda-tensorcore") {
      default_sch_rules = ScheduleRule::DefaultCUDATensorCore();
      default_postprocs = Postproc::DefaultCUDATensorCore();
      default_mutator_probs = Mutator::DefaultCUDATensorCore();
    } else if (kind == "hexagon") {
      default_sch_rules = ScheduleRule::DefaultHexagon();
      default_postprocs = Postproc::DefaultHexagon();
      default_mutator_probs = Mutator::DefaultHexagon();
    } else if (kind == "vnni") {
      default_sch_rules = ScheduleRule::DefaultX86("vnni");
      default_postprocs = Postproc::DefaultCPUTensorization();
      default_mutator_probs = Mutator::DefaultLLVM();
    } else if (kind == "avx512") {
      default_sch_rules = ScheduleRule::DefaultX86("avx512");
      default_postprocs = Postproc::DefaultCPUTensorization();
      default_mutator_probs = Mutator::DefaultLLVM();
    } else if (kind == "asimd") {
      default_sch_rules = ScheduleRule::DefaultARM("neon");
      default_postprocs = Postproc::DefaultCPUTensorization();
      default_mutator_probs = Mutator::DefaultLLVM();
    } else if (kind == "dotprod") {
      default_sch_rules = ScheduleRule::DefaultARM("dotprod");
      default_postprocs = Postproc::DefaultCPUTensorization();
      default_mutator_probs = Mutator::DefaultLLVM();
    } else {
      LOG(FATAL) << "Unsupported kind: " << kind;
      throw;
    }
    if (!sch_rules.defined()) {
      sch_rules = default_sch_rules;
    }
    if (!postprocs.defined()) {
      postprocs = default_postprocs;
    }
    if (!mutator_probs.defined()) {
      mutator_probs = default_mutator_probs;
    }
  }
  if (sch_rules.defined()) {
    for (ScheduleRule i : sch_rules.value()) {
      i->InitializeWithTuneContext(context);
    }
  }
  if (postprocs.defined()) {
    for (Postproc i : postprocs.value()) {
      i->InitializeWithTuneContext(context);
    }
  }
  if (mutator_probs.defined()) {
    for (const auto& kv : mutator_probs.value()) {
      Mutator mutator = kv.first;
      mutator->InitializeWithTuneContext(context);
    }
  }
}

void PySpaceGeneratorNode::InitializeWithTuneContext(const TuneContext& context) {
  ICHECK(f_initialize_with_tune_context != nullptr)
      << "PySpaceGenerator's InitializeWithTuneContext method not implemented!";
  f_initialize_with_tune_context(context);
}

Array<tir::Schedule> PySpaceGeneratorNode::GenerateDesignSpace(const IRModule& mod) {
  ICHECK(f_generate_design_space != nullptr)
      << "PySpaceGenerator's GenerateDesignSpace method not implemented!";
  return f_generate_design_space(mod);
}

SpaceGenerator PySpaceGeneratorNode::Clone() const {
  ICHECK(f_clone != nullptr) << "PySpaceGenerator's Clone method not implemented!";
  return f_clone();
}

SpaceGenerator SpaceGenerator::PySpaceGenerator(
    Optional<Array<ScheduleRule>> sch_rules, Optional<Array<Postproc>> postprocs,
    Optional<Map<Mutator, FloatImm>> mutator_probs,
    FInitializeWithTuneContext f_initialize_with_tune_context,
    FGenerateDesignSpace f_generate_design_space, FClone f_clone) {
  ObjectPtr<PySpaceGeneratorNode> n = make_object<PySpaceGeneratorNode>();
  n->sch_rules = sch_rules;
  n->postprocs = postprocs;
  n->mutator_probs = mutator_probs;
  n->f_initialize_with_tune_context = std::move(f_initialize_with_tune_context);
  n->f_generate_design_space = std::move(f_generate_design_space);
  n->f_clone = std::move(f_clone);
  return SpaceGenerator(n);
}

TVM_REGISTER_OBJECT_TYPE(SpaceGeneratorNode);
TVM_REGISTER_NODE_TYPE(PySpaceGeneratorNode);

TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorInitializeWithTuneContext")
    .set_body_method<SpaceGenerator>(&SpaceGeneratorNode::InitializeWithTuneContext);
TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorGenerateDesignSpace")
    .set_body_method<SpaceGenerator>(&SpaceGeneratorNode::GenerateDesignSpace);
TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorPySpaceGenerator")
    .set_body_typed(SpaceGenerator::PySpaceGenerator);
TVM_REGISTER_GLOBAL("meta_schedule.SpaceGeneratorClone")
    .set_body_method<SpaceGenerator>(&SpaceGeneratorNode::Clone);

}  // namespace meta_schedule
}  // namespace tvm

File: python/tvm/meta_schedule/space_generator/space_generator.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""
Meta Schedule design space generators that generates design
space for generation of measure candidates.
"""
from typing import TYPE_CHECKING, Callable, Dict, List, Optional, Tuple, Union

# isort: off
from typing_extensions import Literal

# isort: on
from tvm._ffi import register_object
from tvm.ir import IRModule
from tvm.runtime import Object
from tvm.tir.schedule import Schedule

from .. import _ffi_api

if TYPE_CHECKING:
    from ..mutator import Mutator
    from ..postproc import Postproc
    from ..schedule_rule import ScheduleRule
    from ..tune_context import TuneContext


@register_object("meta_schedule.SpaceGenerator")
class SpaceGenerator(Object):
    """The abstract design space generator interface."""

    ScheduleFnType = Union[
        Callable[[Schedule], None],  # No output
        Callable[[Schedule], Schedule],  # Single output
        Callable[[Schedule], List[Schedule]],  # Multiple outputs
    ]

    SpaceGeneratorType = Union[
        "SpaceGenerator",
        ScheduleFnType,
        Literal["post-order-apply", "union"],
    ]

    sch_rules: Optional[List["ScheduleRule"]]
    postprocs: Optional[List["Postproc"]]
    mutator_probs: Optional[Dict["Mutator", float]]

    def _initialize_with_tune_context(self, context: "TuneContext") -> None:
        """Initialize the design space generator with tuning context.

        Parameters
        ----------
        context : TuneContext
            The tuning context for initializing the design space generator.
        """
        _ffi_api.SpaceGeneratorInitializeWithTuneContext(  # type: ignore # pylint: disable=no-member
            self, context
        )

    def generate_design_space(self, mod: IRModule) -> List[Schedule]:
        """Generate design spaces given a module.

        Parameters
        ----------
        mod : IRModule
            The module used for design space generation.

        Returns
        -------
        design_spaces : List[Schedule]
            The generated design spaces, i.e., schedules.
        """
        return _ffi_api.SpaceGeneratorGenerateDesignSpace(self, mod)  # type: ignore # pylint: disable=no-member

    def clone(self) -> "SpaceGenerator":
        """Clone the design space generator.

        Returns
        -------
        cloned_sg : SpaceGenerator
            The cloned design space generator.
        """
        return _ffi_api.SpaceGeneratorClone(self)  # type: ignore # pylint: disable=no-member

    @staticmethod
    def create(  # pylint: disable=keyword-arg-before-vararg
        kind: Union[
            Literal["post-order-apply", "union"],
            ScheduleFnType,
        ] = "post-order-apply",
        *args,
        **kwargs,
    ) -> "SpaceGenerator":
        """Create a design space generator."""
        from . import (  # pylint: disable=import-outside-toplevel
            PostOrderApply,
            ScheduleFn,
            SpaceGeneratorUnion,
        )

        if callable(kind):

            def create_schedule_fn(
                func,
                sch_rules=[],
                postprocs=[],
                mutator_probs={},
            ):  # pylint: disable=dangerous-default-value
                return ScheduleFn(func, sch_rules, postprocs, mutator_probs)

            return create_schedule_fn(kind, *args, **kwargs)  # type: ignore
        if kind == "post-order-apply":
            return PostOrderApply(*args, **kwargs)
        if kind == "union":
            return SpaceGeneratorUnion(*args, **kwargs)
        raise ValueError(f"Unknown SpaceGenerator: {kind}")


ScheduleFnType = SpaceGenerator.ScheduleFnType
ScheduleRuleType = Union[
    List["ScheduleRule"],
    Literal["llvm", "cuda", "cuda-tensorcore", "hexagon", "from-target"],
]
PostprocType = Union[
    List["Postproc"],
    Literal["llvm", "cuda", "cuda-tensorcore", "hexagon", "from-target"],
]
MutatorProbType = Union[
    Dict["Mutator", float],
    Literal["llvm", "cuda", "cuda-tensorcore", "hexagon", "from-target"],
]
create = SpaceGenerator.create  # pylint: disable=invalid-name


def _normalize_rules(
    sch_rules: ScheduleRuleType,
    postprocs: PostprocType,
    mutator_probs: MutatorProbType,
) -> Tuple[
    Optional[List["ScheduleRule"]],
    Optional[List["Postproc"]],
    Optional[Dict["Mutator", float]],
]:
    # pylint: disable=import-outside-toplevel
    from ..mutator import Mutator
    from ..postproc import Postproc
    from ..schedule_rule import ScheduleRule

    # pylint: enable=import-outside-toplevel
    assert sch_rules is not None
    assert postprocs is not None
    assert mutator_probs is not None

    if isinstance(sch_rules, str):
        if sch_rules == "from-target":
            sch_rules = None
        else:
            sch_rules = ScheduleRule.create(sch_rules)
    if isinstance(postprocs, str):
        if postprocs == "from-target":
            postprocs = None
        else:
            postprocs = Postproc.create(postprocs)
    if isinstance(mutator_probs, str):
        if mutator_probs == "from-target":
            mutator_probs = None
        else:
            mutator_probs = Mutator.create(mutator_probs)
    return sch_rules, postprocs, mutator_probs  # type: ignore


@register_object("meta_schedule.PySpaceGenerator")
class _PySpaceGenerator(SpaceGenerator):
    """
    A TVM object space generator to support customization on the python side.
    This is NOT the user facing class for function overloading inheritance.

    See also: PySpaceGenerator
    """

    def __init__(
        self,
        sch_rules: ScheduleRuleType = "from-target",
        postprocs: PostprocType = "from-target",
        mutator_probs: MutatorProbType = "from-target",
        f_initialize_with_tune_context: Optional[Callable] = None,
        f_generate_design_space: Optional[Callable] = None,
        f_clone: Optional[Callable] = None,
    ):
        """Constructor."""
        sch_rules, postprocs, mutator_probs = _normalize_rules(sch_rules, postprocs, mutator_probs)

        self.__init_handle_by_constructor__(
            _ffi_api.SpaceGeneratorPySpaceGenerator,  # type: ignore # pylint: disable=no-member
            sch_rules,
            postprocs,
            mutator_probs,
            f_initialize_with_tune_context,
            f_generate_design_space,
            f_clone,
        )


class PySpaceGenerator:
    """
    An abstract space generator with customized methods on the python-side.
    This is the user facing class for function overloading inheritance.

    Note: @derived_object is required for proper usage of any inherited class.
    """

    _tvm_metadata = {
        "cls": _PySpaceGenerator,
        "fields": ["sch_rules", "postprocs", "mutator_probs"],
        "methods": ["_initialize_with_tune_context", "generate_design_space", "clone"],
    }

    def _initialize_with_tune_context(self, context: "TuneContext") -> None:
        """Initialize the design space generator with tuning context.

        Parameters
        ----------
        context : TuneContext
            The tuning context for initializing the design space generator.
        """
        raise NotImplementedError

    def generate_design_space(self, mod: IRModule) -> List[Schedule]:
        """Generate design spaces given a module.

        Parameters
        ----------
        mod : IRModule
            The module used for design space generation.

        Returns
        -------
        design_spaces : List[Schedule]
            The generated design spaces, i.e., schedules.
        """
        raise NotImplementedError

    def clone(self) -> SpaceGenerator:
        """Clone the design space generator.

        Returns
        -------
        cloned_sg : SpaceGenerator
            The cloned design space generator.
        """
        raise NotImplementedError
